test/distributed/optim/test_zero_redundancy_optimizer.py

```diff
@@ -656,6 +657,113 @@ def closure_sharded(input_tensor=input_tensor):
+            for _ in range(NUM_EPOCHS):
+                for input in inputs:
+                    output = ddp_model(input)
+                    loss_fn(output, labels).backward()
+                    if rank == world_size - 1:
+                        grads = []
+                        for p in ddp_model.parameters():
+                            grads.append(p.grad.detach().clone().to(self.device))
+                    local_optim.step()
+                    if rank == world_size - 1:
+                        params = []
+                        for p in ddp_model.parameters():
+                            params.append(p.detach().clone().to(self.device))
+                        grads_at_each_iter.append(grads)
+                        params_at_each_iter.append(params)
+
+        # Broadcast the gradients and parameters to all ranks
+        grads_and_params = [grads_at_each_iter, params_at_each_iter]
+        grads_and_params = _broadcast_object(grads_and_params, src_rank=world_size - 1, group=dist.group.WORLD, device=self.device)
+        grads_at_each_iter = grads_and_params[0]
+        params_at_each_iter = grads_and_params[1]
+        # TODO: Replace this `_broadcast_object` with `broadcast_object_list`
+        # once the latter supports loading to the destination device instead
+        # of the source device
+
+        # A process must still set the remaining gradients after joining, so we
+        # define a join hook to do this before the ZeRO join hook
+        class _JoinGradInfo():
+            def __init__(self, grads, device):
+                self.grads = grads  # remaining gradients to set (in order)
+                self.index = 0
+                self.device = device
+
+        def set_grads_join_hook(zero_optim, grads, device):
+            def pre_hook(zero_optim):
+                """Saves information needed for the main hook."""
+                zero_optim._join_grad_info = _JoinGradInfo(grads, device)
+
+            def main_hook(zero_optim):
+                """Sets the gradients manually."""
+                grads = zero_optim._join_grad_info.grads[zero_optim._join_grad_info.index]
+                zero_optim._join_grad_info.index += 1
+                for p, grad in zip(zero_optim._all_params, grads):
+                    p.grad = grad.detach().clone().to(zero_optim._join_grad_info.device)
+
+            return _JoinHook(zero_optim, pre_hook, main_hook, None)
+
+        num_grads_after_joining = NUM_EPOCHS * (world_size - rank - 1)
+        grads = grads_at_each_iter[-num_grads_after_joining:]
+        set_grads_jh = set_grads_join_hook(zero_optim, grads, self.device)
+        zero_jh = zero_optim._join_hook()
+        iter = 0
+        with _Join([set_grads_jh, zero_jh]):
+            for _ in range(NUM_EPOCHS):
+                for input in inputs:
+                    # Schedule an all-reduce to indicate not joined
+                    dist.all_reduce(torch.ones(1, device=self.device), group=dist.group.WORLD)
```

Comment:
Yes, I was thinking we require a contract between the programmer and join context manager: The programmer must ensure that somewhere there is an all-reduce from non-joined processes with value `1` to signal to the join context manager that those processes have not joined yet. For DDP, this is in `forward()`. In the case of this test, since there is no DDP, I included the all-reduce explicitly. If the DDP and ZeRO join hooks are used together, there is no need for this explicit all-reduce.